flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java

```diff
@@ -0,0 +1,345 @@
+import io.netty.util.CharsetUtil;
+import io.netty.util.concurrent.GenericFutureListener;
+import org.apache.flink.runtime.blob.BlobKey;
+import org.apache.flink.runtime.instance.ActorGateway;
+import org.apache.flink.runtime.instance.Instance;
+import org.apache.flink.runtime.instance.InstanceID;
+import org.apache.flink.runtime.messages.JobManagerMessages;
+import org.apache.flink.runtime.messages.TaskManagerMessages;
+import org.apache.flink.runtime.webmonitor.JobManagerRetriever;
+import org.apache.flink.runtime.webmonitor.files.MimeTypes;
+import org.apache.flink.runtime.webmonitor.files.StaticFileServerHandler;
+import org.apache.flink.util.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Option;
+import scala.Tuple2;
+import scala.concurrent.Await;
+import scala.concurrent.ExecutionContextExecutor;
+import scala.concurrent.Future;
+import scala.concurrent.duration.FiniteDuration;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.netty.handler.codec.http.HttpHeaders.Names.CONNECTION;
+import static io.netty.handler.codec.http.HttpHeaders.Names.CONTENT_TYPE;
+import static io.netty.handler.codec.http.HttpResponseStatus.INTERNAL_SERVER_ERROR;
+import static io.netty.handler.codec.http.HttpResponseStatus.NOT_FOUND;
+import static io.netty.handler.codec.http.HttpResponseStatus.OK;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+/**
+ * Request handler that returns the TaskManager log/out files.
+ *
+ * <p>This code is based on the "HttpStaticFileServerHandler" from the Netty project's HTTP server
+ * example.</p>
+ */
+@ChannelHandler.Sharable
+public class TaskManagerLogHandler extends SimpleChannelInboundHandler<Routed> {
+
+	/** Default logger, if none is specified */
+	private static final Logger DEFAULT_LOGGER = LoggerFactory.getLogger(StaticFileServerHandler.class);
+
+	// ------------------------------------------------------------------------
+
+	/** JobManager retriever */
+	private final JobManagerRetriever retriever;
+
+	private final Future<String> localJobManagerAddressFuture;
+
+	private final FiniteDuration timeout;
+
+	/** Keep track of last transmitted log, to clean up old ones */
+	private final HashMap<String, BlobKey> lastSubmittedLog = new HashMap<>();
+	private final HashMap<String, BlobKey> lastSubmittedStdout = new HashMap<>();
+
+	/** Keep track of request status, prevents multiple log requests for a single TM running concurrently */
+	private final HashMap<String, Boolean> lastRequestCompleted = new HashMap<>();
+
+	/** The log for all error reporting */
+	private final Logger logger;
+
+	/** indicates which log file should be displayed; true indicates .log, false indicates .out */
+	private final boolean serveLogFile;
```

Comment:
i think it's overkill for a binary option, and won't make the code more readable.